fs/vfs/fs_eventfd.c

```diff
@@ -0,0 +1,623 @@
+static ssize_t eventfd_do_write(FAR struct file *filep,
+                                FAR const char *buffer, size_t len);
+static int eventfd_do_ioctl(FAR struct file *filep, int cmd,
+                            unsigned long arg);
+#ifdef CONFIG_EVENT_FD_POLL
+static int eventfd_poll(FAR struct file *filep, FAR struct pollfd *fds,
+                       bool setup);
+#endif
+
+static int eventfd_alloc_dev(eventfd_t counter, int flags);
+
+#ifdef CONFIG_EVENT_FD_POLL
+static void eventfd_pollnotify(FAR struct eventfd_priv_s *dev,
+                               pollevent_t eventset);
+#endif
+
+static int get_unique_minor(void);
+static void release_minor(int minor);
+
+static FAR struct eventfd_priv_s *eventfd_allocdev(void);
+static void eventfd_destroy(FAR struct eventfd_priv_s *dev);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static const struct file_operations g_eventfd_fops =
+{
+  eventfd_open,     /* open */
+  eventfd_close,    /* close */
+  eventfd_do_read,  /* read */
+  eventfd_do_write, /* write */
+  0,                /* seek */
+  eventfd_do_ioctl  /* ioctl */
+#ifdef CONFIG_EVENT_FD_POLL
+  , eventfd_poll    /* poll */
+#endif
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+FAR struct eventfd_priv_s *eventfd_allocdev(void)
+{
+  FAR struct eventfd_priv_s *dev;
+
+  dev = (FAR struct eventfd_priv_s *)
+    kmm_malloc(sizeof(struct eventfd_priv_s));
+  if (dev)
+    {
+      /* Initialize the private structure */
+
+      memset(dev, 0, sizeof(struct eventfd_priv_s));
+      nxsem_init(&dev->exclsem, 0, 0);
+      nxsem_init(&dev->rdsem, 0, 0);
+      nxsem_init(&dev->wrsem, 0, 0);
+
+      /* The read/write wait semaphores are used for signaling
+       * and should not have priority inheritance enabled.
+       */
+
+      nxsem_set_protocol(&dev->rdsem, SEM_PRIO_NONE);
+      nxsem_set_protocol(&dev->wrsem, SEM_PRIO_NONE);
+    }
+
+  return dev;
+}
+
+void eventfd_destroy(FAR struct eventfd_priv_s *dev)
+{
+  nxsem_destroy(&dev->exclsem);
+  nxsem_destroy(&dev->rdsem);
+  nxsem_destroy(&dev->wrsem);
+  kmm_free(dev);
+}
+
+#ifdef CONFIG_EVENT_FD_POLL
+void eventfd_pollnotify(FAR struct eventfd_priv_s *dev,
+                               pollevent_t eventset)
+{
+  FAR struct pollfd *fds;
+  int i;
+
+  for (i = 0; i < CONFIG_EVENT_FD_NPOLLWAITERS; i++)
+    {
+      fds = dev->d_fds[i];
+      if (fds)
+        {
+          fds->revents |= eventset & fds->events;
+
+          if (fds->revents != 0)
+            {
+              nxsem_post(fds->sem);
+            }
+        }
+    }
+}
+#endif
+
+int get_unique_minor(void)
+{
+  static int minor = 0;
+
+  /* TODO reimplement this with minor bit map ? */
+
+  return (minor++) & 511;
```

Comment:
why mask 511?

Reply:
I guess 512 limit is enough. This limit is done to avoid dealing with long device name, ie. "/dev/efd8471987393874"

Reply:
bitmap implementation should be much better than this counter. It will no longer work properly after a 511 overflow...

Reply:
since the minor in eventfd_priv_s is uint8_t, it si better to use 0xff.